### PR TITLE
add ncli_db subcommand to prune database of unnecessary blocks and states

### PR DESCRIPTION
**Pull request by tersec**
```
db:
total 1.1G
-rw-r--r-- 1 1021M Aug 31 13:51 nbc.sqlite3
-rw-r--r-- 1   32K Aug 31 13:51 nbc.sqlite3-shm
-rw-r--r-- 1  7.2M Aug 31 13:51 nbc.sqlite3-wal

db_unpruned/:
total 51G
-rw-r--r-- 1  50G Aug 31 13:46 nbc.sqlite3
-rw-r--r-- 1 608K Aug 31 13:46 nbc.sqlite3-shm
-rw-r--r-- 1 291M Aug 31 13:46 nbc.sqlite3-wal
```

### CODE REVIEW

**Comment by stefantalpalaru** [comment]
How are we integrating this into NBC?

**Comment by tersec** [comment]
Separately. This is an offline algorithm, which takes several minutes for me to run.

I anticipate using a different, more incremental algorithm for nbc.

**Comment by arnetheduck** [comment]
worth keeping in mind that a database like this will not be usable for command line fetching of arbitrary states

**Comment by tersec** [comment]
The other part that I didn't emphasize as a design choice, maybe, is that nowhere in the spec nor the nbc code does it refer to states (not blocks) before the finalized state, except for a few sanity/integrity checks which amount to checking for the existence of a tail state. This doesn't copy those, but they can be relatively cheaply reconstructed from the tail state and the blocks, if one wants to sort of create a full-archive database.

This also means that if one wants to re-integrity-check the full database, that's no harder (start at tail state and apply the blocks), but checking for any given block's state/block/state-sequence's validity, or any block's validity, means reconstructing the state before it, which by induction/recursion goes back to the tail state, if one's before the finalized state. This is fine for just running a node, and helps quite a bit in reducing storage usage, but it's possible there are use cases where one wants actual random-access to re-run state transitions using pre-finalization blocks.

**Comment by arnetheduck** [comment]
well, the spec does reference the genesis state (which the tail state presently always matches, but won't in the future)

**Comment by tersec** [comment]
As true as that is, there's therefore not really a separate representation in the database to keep. Adding a special case akin to the tail state, for a genesis state, is a constant-time addition that just doesn't make sense yet.

I looked, separately, at just not reconstructing the block DAG in any form in nbc older than the finalization epoch, but that takes the opportunity to read in the whole chain, as here, more or less for the purpose of checking that there's an intact, continuous head <-> tail linkage, which then doesn't end up mattering that I can discern. As the chain grows, but finalization delay doesn't, that's going to end up more and more inefficient, htr or SSZ or etc loading efficiencies aside. Just ignore the pre-finalization blocks until they're requested by `getBlocksByRange`, etc.

So obviously this all needs some changes, but this PR works with the database schema we have.

Meanwhile, https://github.com/status-im/nim-beacon-chain/pull/1593/commits/501af302a538c4cc00e7238844d0916f60ca2428 keeps pre-finalization states by default, but (1) is significantly slower (start-to-finish was about 5 minutes for me; it's been at least that long, and isn't even half-done, from the same starting un-pruned database; and (2) is already 7.5GB, probably ending up at 10 or 15GB at least. Better than the starting 50GB, but a lot of that's an artifact of the long non-finalization period resulting in pruned DAG branches. I'm not convinced this is a good default for most non-developer users.

To the degree there's some API which depends on that, just as with Geth or similar, it might be worth supporting that sort of full-archive type mode, but it's just so costly, to the point of impracticality for our nominally ultimately targeted platforms absent much cleverer encoding.

**Comment by tersec** [comment]
When keeping old states:
```
$ rm build/data/shared_medalla_0/db/nbc_pruned.sqlite3* -fv && time build/ncli_db pruneDatabase --verbose --db=build/data/shared_medalla_0/db/ > prune_log.txt

real	39m34.169s
user	27m43.730s
sys	4m43.097s
```
So about 8x as slow.

And the result is 20x as large (though still 40% of the size of the original; same starting point as for the 1GB pruned database):
```
-rw-r--r-- 1  19G Aug 31 17:37 nbc_pruned.sqlite3
-rw-r--r-- 1  32K Aug 31 17:37 nbc_pruned.sqlite3-shm
-rw-r--r-- 1 7.4M Aug 31 17:37 nbc_pruned.sqlite3-wal
```

File: beacon_chain/beacon_chain_db.nim [hunk 1]
```diff
@@ -219,3 +219,73 @@ iterator getAncestors*(db: BeaconChainDB, root: Eth2Digest):
   while db.get(subkey(SignedBeaconBlock, res.root), res) == GetResult.found:
     yield res
     res.root = res.message.parent_root
+
+proc copyPrunedDatabase*(
```
Comment (arnetheduck) [inline beacon_chain/beacon_chain_db.nim:223]: this pretty much uses the public api and is only used in ncli - I don't think we'll use it elsewhere as-is, so perhaps keep in ncli_db for now?

Comment (tersec) [reply beacon_chain/beacon_chain_db.nim]: https://github.com/status-im/nim-beacon-chain/pull/1593/commits/86a32ac9008672801f8439acb5d2a8dbfda216a1

Comment (tersec) [reply beacon_chain/beacon_chain_db.nim]: The usage of the public API was one reason I chose this design to begin with -- it avoided what would have been cascading changes in `nim-eth`, having to potentially create some compatibility shim in the rocksdb k/v layer, et cetera. Secondarily, it more naturally implements the safer create-temporary-file-then-atomic-rename pattern than the delete-complement-of-desired-key approach.

**Pull request merged**